lib/deimos/active_record/batch_consumption.rb

```diff
@@ -0,0 +1,133 @@
+# frozen_string_literal: true
+
+require 'deimos/active_record/batch_slicer'
+require 'deimos/utils/deadlock_retry'
+require 'deimos/message'
+
+module Deimos
+  module ActiveRecord
+    # Methods for consuming batches of messages and saving them to the database
+    # in bulk ActiveRecord operations.
+    module BatchConsumption
+      # Handle a batch of Kafka messages. Batches are split into "slices",
+      # which are groups of independent messages that can be processed together
+      # in a single database operation.
+      # If two messages in a batch have the same key, we cannot process them
+      # in the same operation as they would interfere with each other. Thus
+      # they are split
+      # @param payloads [Array<Hash>] Decoded payloads.
+      # @param metadata [Hash] Information about batch, including keys.
+      def consume_batch(payloads, metadata)
+        messages = payloads.
+          zip(metadata[:keys]).
+          map { |p, k| Deimos::Message.new(p, nil, key: k) }
```

Comment:
OK - now I understand what the slicing is supposed to achieve. I'm not convinced this whole slicing thing is actually necessary. If the records have an ID, there is literally no point in us saving them three or four times in a single batch. The last value will "win". The only possible reason we could have for updating it in different slices would be if there was some kind of database trigger that happened on update.

I'd prefer to have compaction turned on by default, and have the user turn it off if and only if the table has no primary key, or they know that the topic is insert-only (i.e. there is no primary key, or the primary key is not in the topic).

Reply:
I'm OK with making compacted the default. Will make those changes

There are other situations where it makes sense to have it not be compacted. The consumer may do some custom handling where they need to add some app logic for every event. It is also useful for when the topic represents something like an audit log and you use soft deletion. If you have an entity that gets created and then destroyed immediately after, you would still want that record to appear in your DB with a deleted flag set. We do something similar to this in the item-hippocampus

Reply:
You've already got the work done, so I'm OK leaving it in. Just not super stoked about having to maintain this when the use case seems to be pretty obscure. Is there some way we can separate it out a bit more (e.g. having a different subclass that handles this as opposed to the "paved road" where it's compacted)?

Reply:
Do you think a subclass of ActiveRecordConsumer would be clean enough? The other alternative I can think of is another module you include to override `consume_batch` but that seems uglier

Reply:
I think you can have two different methods at least, one that handles slicing and one that doesn't, and you call them differently depending on settings. The methods could be in separate modules but that might be overkill.